napari/experimental/_progressive_loading.py

```diff
@@ -0,0 +1,1652 @@
+            rendering=rendering,
+            contrast_limits=contrast_limits,
+        )
+        layers[scale] = layer
+        layer.metadata["translated"] = False
+
+    # Linked list of layers for visibility control
+    for scale in reversed(range(len(layers))):
+        layers[scale].metadata["prev_layer"] = (
+            layers[scale + 1]
+            if scale < len(multiscale_data._data) - 1
+            else None
+        )
+
+    top_left = canvas_corners[0, :]
+    bottom_right = canvas_corners[1, :]
+    LOGGER.debug(f'>>> top left: {top_left}, bottom_right: {bottom_right}')
+    LOGGER.info(f"viewer canvas corners {canvas_corners}")
+
+    # Connect to camera and dims
+    for listener in [viewer.camera.events, viewer.dims.events]:
+        listener.connect(
+            debounced(
+                ensure_main_thread(
+                    dims_update_handler(
+                        data=multiscale_data,
+                        viewer=viewer,
+                        ndisplay=ndisplay,
+                    )
+                ),
+                timeout=2000,
+            )
+        )
+
+    # Trigger first render
+    dims_update_handler(
+        None, data=multiscale_data, viewer=viewer, ndisplay=ndisplay
+    )
+
+    return viewer
+
+
+# ---------- 2D specific ----------
+
+
+def chunk_priority_2D(chunk_keys, corner_pixels, scale_factor):
+    """Return the keys for all chunks at this scale within the corner_pixels.
+
+    Parameters
+    ----------
+    chunk_keys : list of list of slices for each dimension
+        a list of list of slices for each dimension
+    corner_pixels : tuple
+        ND top left and bottom right coordinates for the current view
+    scale_factor : float
+        the scale factor for this scale level
+
+    """
+    mins = corner_pixels[0, :] / scale_factor
+    maxs = corner_pixels[1, :] / scale_factor
+
+    contained_keys = chunk_keys_within_interval(chunk_keys, mins, maxs)
+
+    priority_map: List = []
+
+    for _idx, chunk_key in enumerate(
+        list(
+            itertools.product(
+                *[contained_keys[k] for k in sorted(contained_keys.keys())]
+            )
+        )
+    ):
+        priority = 0
+        # TODO filter priority here
+        priority = 0 if True else np.inf
+        if priority < np.inf:
+            heapq.heappush(priority_map, (priority, chunk_key))
+
+    return priority_map
+
+
+def should_render_scale_2D(scale, viewer, min_scale, max_scale):
+    """Test if a scale should be rendered.
+
+    Parameters
+    ----------
+    scale : int
+        a scale level
+    viewer : napari.viewer.Viewer
+        a napari viewer
+    min_scale : int
+        the minimum scale level to show
+    max_scale : int
+        the maximum scale level to show
+    """
+    layer_name = get_layer_name_for_scale(scale)
+    layer = viewer.layers[layer_name]
+    layer_scale = layer.scale
+
+    pixel_size = viewer.camera.zoom * max(layer_scale)
+
+    if max_scale == 7:
```

Comment:
What's special about scale 7?

Reply:
This was definitely a hack that we needed earlier. This part of the code needs to be revisited.